exchanges/orderbook/orderbook.go

```diff
@@ -44,6 +45,12 @@ func SubscribeToExchangeOrderbooks(exchange string) (dispatch.Pipe, error) {
 // Update stores orderbook data
 func (s *Service) Update(b *Base) error {
 	name := strings.ToLower(b.Exchange)
+	mapKey := key.PairAssetKey{
```

Comment:
Out of scope Suggestion (This PR or I can open another PR later): Had a quick check out and we can squish all this stuff into one by using a temp lookup slice. I stuffed up the ID's for the exchange and the specific book. But you get the gist. 😆 **Can also consider doing the same with accounts**

```diff
diff --git a/exchanges/orderbook/orderbook.go b/exchanges/orderbook/orderbook.go
index 983d1a210..cd3f149af 100644
--- a/exchanges/orderbook/orderbook.go
+++ b/exchanges/orderbook/orderbook.go
@@ -32,57 +32,53 @@ func DeployDepth(exchange string, p currency.Pair, a asset.Item) (*Depth, error)

 // SubscribeToExchangeOrderbooks returns a pipe to an exchange feed
 func SubscribeToExchangeOrderbooks(exchange string) (dispatch.Pipe, error) {
+       exchange = strings.ToLower(exchange)
        service.mu.Lock()
        defer service.mu.Unlock()
-       exch, ok := service.books[strings.ToLower(exchange)]
-       if !ok {
-               return dispatch.Pipe{}, fmt.Errorf("%w for %s exchange",
-                       errCannotFindOrderbook, exchange)
+       for x := range service.ids {
+               if service.ids[x].Exchange == exchange {
+                       return service.Mux.Subscribe(service.ids[x].ID)
+               }
        }
-       return service.Mux.Subscribe(exch.ID)
+       return dispatch.Pipe{}, fmt.Errorf("%w for %s exchange",
+               errCannotFindOrderbook, exchange)
 }

 // Update stores orderbook data
 func (s *Service) Update(b *Base) error {
        name := strings.ToLower(b.Exchange)
-       mapKey := key.PairAssetKey{
-               Base:  b.Pair.Base.Item,
-               Quote: b.Pair.Quote.Item,
-               Asset: b.Asset,
+       mapKey := key.ExchangePairAssetKey{
+               Exchange: name,
+               Base:     b.Pair.Base.Item,
+               Quote:    b.Pair.Quote.Item,
+               Asset:    b.Asset,
        }

        s.mu.Lock()
-       m1, ok := s.books[name]
+       depth, ok := s.books[mapKey]
        if !ok {
-               id, err := s.Mux.GetID()
+               id, err := s.muxIDForExchange(name)
                if err != nil {
                        s.mu.Unlock()
                        return err
                }
-               m1 = Exchange{
-                       m:  make(map[key.PairAssetKey]*Depth),
-                       ID: id,
-               }
-               s.books[name] = m1
-       }
-       book, ok := m1.m[mapKey]
-       if !ok {
-               book = NewDepth(m1.ID)
-               book.AssignOptions(b)
-               m1.m[mapKey] = book
+               depth = NewDepth(id)
+               depth.AssignOptions(b)
+               s.books[mapKey] = depth
        }
-       err := book.LoadSnapshot(b.Bids, b.Asks, b.LastUpdateID, b.LastUpdated, true)
+
+       err := depth.LoadSnapshot(b.Bids, b.Asks, b.LastUpdateID, b.LastUpdated, true)
        s.mu.Unlock()
        if err != nil {
                return err
        }
-       return s.Mux.Publish(book, m1.ID)
+       return s.Mux.Publish(depth, depth._ID)
 }

 // DeployDepth used for subsystem deployment creates a depth item in the struct
 // then returns a ptr to that Depth item
-func (s *Service) DeployDepth(exchange string, p currency.Pair, a asset.Item) (*Depth, error) {
-       if exchange == "" {
+func (s *Service) DeployDepth(name string, p currency.Pair, a asset.Item) (*Depth, error) {
+       if name == "" {
                return nil, errExchangeNameUnset
        }
        if p.IsEmpty() {
@@ -91,64 +87,55 @@ func (s *Service) DeployDepth(exchange string, p currency.Pair, a asset.Item) (*
        if !a.IsValid() {
                return nil, errAssetTypeNotSet
        }
-       mapKey := key.PairAssetKey{
-               Base:  p.Base.Item,
-               Quote: p.Quote.Item,
-               Asset: a,
+       name = strings.ToLower(name)
+       mapKey := key.ExchangePairAssetKey{
+               Exchange: name,
+               Base:     p.Base.Item,
+               Quote:    p.Quote.Item,
+               Asset:    a,
        }

        s.mu.Lock()
        defer s.mu.Unlock()
-       m1, ok := s.books[strings.ToLower(exchange)]
+       depth, ok := s.books[mapKey]
        if !ok {
-               id, err := s.Mux.GetID()
+               id, err := s.muxIDForExchange(name)
                if err != nil {
+                       s.mu.Unlock()
                        return nil, err
                }
-               m1 = Exchange{
-                       m:  make(map[key.PairAssetKey]*Depth),
-                       ID: id,
-               }
-               s.books[strings.ToLower(exchange)] = m1
-       }
-       book, ok := m1.m[mapKey]
-       if !ok {
-               book = NewDepth(m1.ID)
-               book.exchange = exchange
-               book.pair = p
-               book.asset = a
-               m1.m[mapKey] = book
+
+               depth = NewDepth(id)
+               depth.pair = p
+               depth.asset = a
+               depth.exchange = name
+               s.books[mapKey] = depth
        }
-       return book, nil
+       return depth, nil
 }

 // GetDepth returns the actual depth struct for potential subsystems and
 // strategies to interact with
-func (s *Service) GetDepth(exchange string, p currency.Pair, a asset.Item) (*Depth, error) {
+func (s *Service) GetDepth(name string, p currency.Pair, a asset.Item) (*Depth, error) {
        s.mu.Lock()
        defer s.mu.Unlock()
-       m1, ok := s.books[strings.ToLower(exchange)]
-       if !ok {
-               return nil, fmt.Errorf("%w for %s exchange",
-                       errCannotFindOrderbook, exchange)
-       }
-
-       book, ok := m1.m[key.PairAssetKey{
-               Base:  p.Base.Item,
-               Quote: p.Quote.Item,
-               Asset: a,
+       depth, ok := s.books[key.ExchangePairAssetKey{
+               Exchange: name,
+               Base:     p.Base.Item,
+               Quote:    p.Quote.Item,
+               Asset:    a,
        }]
        if !ok {
-               return nil, fmt.Errorf("%w associated with base currency %s",
+               return nil, fmt.Errorf("%w associated for %v %v",
                        errCannotFindOrderbook,
-                       p.Quote)
+                       p, a)
        }
-       return book, nil
+       return depth, nil
 }

 // Retrieve gets orderbook depth data from the associated linked list and
 // returns the base equivalent copy
-func (s *Service) Retrieve(exchange string, p currency.Pair, a asset.Item) (*Base, error) {
+func (s *Service) Retrieve(name string, p currency.Pair, a asset.Item) (*Base, error) {
        if p.IsEmpty() {
                return nil, currency.ErrCurrencyPairEmpty
        }
@@ -156,25 +143,22 @@ func (s *Service) Retrieve(exchange string, p currency.Pair, a asset.Item) (*Bas
                return nil, fmt.Errorf("%w %v", asset.ErrNotSupported, a)
        }

+       name = strings.ToLower(name)
+
        s.mu.Lock()
        defer s.mu.Unlock()
-       m1, ok := s.books[strings.ToLower(exchange)]
-       if !ok {
-               return nil, fmt.Errorf("%w for %s exchange",
-                       errCannotFindOrderbook,
-                       exchange)
-       }
-       book, ok := m1.m[key.PairAssetKey{
-               Base:  p.Base.Item,
-               Quote: p.Quote.Item,
-               Asset: a,
+       depth, ok := s.books[key.ExchangePairAssetKey{
+               Exchange: name,
+               Base:     p.Base.Item,
+               Quote:    p.Quote.Item,
+               Asset:    a,
        }]
        if !ok {
                return nil, fmt.Errorf("%w associated with base currency %s",
                        errCannotFindOrderbook,
                        p.Quote)
        }
-       return book.Retrieve()
+       return depth.Retrieve()
 }

 // GetDepth returns the concrete book allowing the caller to stream orderbook changes
diff --git a/exchanges/orderbook/orderbook_types.go b/exchanges/orderbook/orderbook_types.go
index a60360af2..be4b937cb 100644
--- a/exchanges/orderbook/orderbook_types.go
+++ b/exchanges/orderbook/orderbook_types.go
@@ -36,22 +36,35 @@ var (
 )

 var service = Service{
-       books: make(map[string]Exchange),
+       books: make(map[key.ExchangePairAssetKey]*Depth),
        Mux:   dispatch.GetNewMux(nil),
 }

+type MuxID struct {
+       Exchange string
+       ID       uuid.UUID
+}
+
 // Service provides a store for difference exchange orderbooks
 type Service struct {
-       books map[string]Exchange
+       books map[key.ExchangePairAssetKey]*Depth
+       ids   []MuxID
        *dispatch.Mux
        mu sync.Mutex
 }

-// Exchange defines a holder for the exchange specific depth items with a
-// specific ID associated with that exchange
-type Exchange struct {
-       m  map[key.PairAssetKey]*Depth
-       ID uuid.UUID
+func (s *Service) muxIDForExchange(exch string) (uuid.UUID, error) {
+       for x := range s.ids {
+               if s.ids[x].Exchange == exch {
+                       return s.ids[x].ID, nil
+               }
+       }
+       id, err := s.Mux.GetID()
+       if err != nil {
+               return uuid.Nil, err
+       }
+       s.ids = append(s.ids, MuxID{Exchange: exch, ID: id})
+       return id, nil
 }

 // Item stores the amount and price values
```

Reply:
This is a bit to change the design of multiple systems and ensure it all works. I'll do it, but this will be delayed now

Reply:
Doesn't need to be done, happy with a defer.

Reply:
Yeah I think defer is best for this